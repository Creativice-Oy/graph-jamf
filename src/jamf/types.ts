import {
  Admin,
  AdminsAndGroups,
  ApplicationDetail,
  Computer,
  ComputerDetail,
  Configuration,
  ConfigurationDetail,
  Group,
  MobileDevice,
  User,
} from "../types";

export interface AccountsResponse {
  accounts: AdminsAndGroups;
}

export interface AdminResponse {
  account: Admin;
}

export interface GroupResponse {
  group: Group;
}

export interface UsersResponse {
  users: User[];
}

export interface UserResponse {
  user: User;
}

export interface ApplicationDetailResponse {
  computer_applications: ApplicationDetail;
}

export interface ComputerResponse {
  computers: Computer[];
}

export interface ComputerDetailResponse {
  computer: ComputerDetail;
}

export interface ConfigurationResponse {
  computer_configurations: Configuration[];
}

export interface ConfigurationDetailResponse {
  computer_configuration: ConfigurationDetail;
}

export interface MobileDevicesResponse {
  mobile_devices: MobileDevice[];
}

export interface JamfDataModel {
  users: User[];
  mobileDevices: MobileDevice[];
  computers: Computer[];
  computerDetails: ComputerDetail[];
  configurationDetails: ConfigurationDetail[];
  admins: Admin[];
  groups: Group[];
}
