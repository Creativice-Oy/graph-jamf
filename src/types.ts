import {
  GraphClient,
  IntegrationExecutionContext,
  PersisterClient,
} from "@jupiterone/jupiter-managed-integration-sdk";

import JamfClient from "./jamf/JamfClient";

export interface JamfIntegrationContext extends IntegrationExecutionContext {
  graph: GraphClient;
  persister: PersisterClient;
  provider: JamfClient;
  account: Account;
}

export interface Account {
  id: string;
  name: string;
}

export interface AdminsAndGroups {
  users: Admin[];
  groups: Group[];
}

export interface Admin {
  id: number;
  name: string;
  directory_user?: boolean;
  full_name?: string;
  email?: string;
  email_address?: string;
  password_sha256?: string;
  enabled?: string;
  force_password_change?: boolean;
  access_level?: string;
  privilege_set?: string;
  privileges?: Privileges;
  groups?: {
    group: {
      id: number;
      name: string;
      site: { id: number; name: string };
      privileges: {
        jss_objects?: any;
        jss_settings?: any;
        recon?: any;
        casper_admin?: any;
        casper_remote?: any;
        casper_imaging?: any;
      };
    };
  };
}

export interface Group {
  id: number;
  name: string;
  site: { id: number; name: string };
  access_level?: string;
  privilege_set?: string;
  privileges?: Privileges;
  members?: Array<{ id: number; name: string }>;
}

export interface Privileges {
  jss_objects: string[];
  jss_settings?: string[];
  jss_actions?: string[];
  recon: string[];
  casper_admin?: string[];
  casper_remote: string[];
  casper_imaging: string[];
}

export interface User {
  id: number;
  name: string;
  full_name?: string;
  email?: string;
  email_address?: string;
  phone_number?: string;
  position?: string;
  enable_custom_photo_url?: boolean;
  custom_photo_url?: string;
  ldap_server?: {
    id: number;
    name: string;
  };
  extension_attributes?: Array<{
    extension_attribute: {
      id: number;
      name: string;
      type: string;
      value: string;
    };
  }>;
  sites?: Array<{
    site: {
      id: number;
      name: string;
    };
  }>;
  links?: {
    computers: Array<{
      id: number;
      name: string;
    }>;
    peripherals: Array<{
      id: number;
      name: string;
    }>;
    mobile_devices: Array<{
      id: number;
      name: string;
    }>;
    vpp_assignments: Array<{
      id: number;
      name: string;
    }>;
    total_vpp_code_count: number;
  };
}

export interface Computer {
  id: number;
  name: string;
  managed: boolean;
  username: string;
  model: string;
  department: string;
  building: string;
  mac_address: string;
  udid: string;
  serial_number: string;
  report_date_utc: string;
  report_date_epoch: number;
}

export interface Application {
  name: string;
  path: string;
  version: string;
}

export interface ApplicationComputer {
  id: number;
  name: string;
  udid: string;
  serial_number: string;
  mac_address: string;
}

export interface ApplicationDetail {
  versions: Array<{
    number: string;
    computers: ApplicationComputer[];
  }>;
  unique_computers: ApplicationComputer[];
}

export interface Storage {
  disk: string;
  model: string;
  revision: string;
  serial_number: string;
  size: number;
  drive_capacity_mb: number;
  connection_type: string;
  smart_status: string;
  partition: {
    name: string;
    size: number;
    type: string;
    partition_capacity_mb: number;
    percentage_full: number;
    filevault_status: string;
    filevault_percent: number;
    filevault2_status: string;
    filevault2_percent: number;
    boot_drive_available_mb: number;
    lvgUUID: string;
    lvUUID: string;
    pvUUID: string;
  };
}

export interface Printer {
  name: string;
  uri: string;
  type: string;
  location: string;
}

export interface ComputerDetail {
  general: {
    id: number;
    name: string;
    mac_address: string;
    alt_mac_address: string;
    ip_address: string;
    last_reported_ip: string;
    serial_number: string;
    udid: string;
    jamf_version: string;
    platform: string;
    barcode_1: string;
    barcode_2: string;
    asset_tag: string;
    remote_management: {
      managed: boolean;
      management_username: string;
      management_password_sha256: string;
    };
    mdm_capable: boolean;
    mdm_capable_users: { mdm_capable_user: string };
    management_status: {
      enrolled_via_dep: boolean;
      user_approved_enrollment: boolean;
      user_approved_mdm: boolean;
    };
    report_date: string;
    report_date_epoch: number;
    report_date_utc: string;
    last_contact_time: string;
    last_contact_time_epoch: number;
    last_contact_time_utc: string;
    initial_entry_date: string;
    initial_entry_date_epoch: number;
    initial_entry_date_utc: string;
    last_cloud_backup_date_epoch: 0;
    last_cloud_backup_date_utc: string;
    last_enrolled_date_epoch: number;
    last_enrolled_date_utc: string;
    distribution_point: string;
    sus: string;
    netboot_server: string;
    site: { id: number; name: string };
    itunes_store_account_is_active: boolean;
  };
  software: {
    unix_executables: any[];
    licensed_software: any[];
    installed_by_casper: any[];
    installed_by_installer_swu: string[];
    cached_by_casper: any[];
    available_software_updates: any[];
    available_updates: any;
    running_services: string[];
    applications: Application[];
    fonts: any[];
    plugins: any[];
  };
  hardware: {
    make: string;
    model: string;
    model_identifier: string;
    os_name: string;
    os_version: string;
    os_build: string;
    master_password_set: boolean;
    active_directory_status: string;
    service_pack: string;
    processor_type: string;
    processor_architecture: string;
    processor_speed: number;
    processor_speed_mhz: number;
    number_processors: number;
    number_cores: number;
    total_ram: number;
    total_ram_mb: number;
    boot_rom: string;
    bus_speed: number;
    bus_speed_mhz: number;
    battery_capacity: number;
    cache_size: number;
    cache_size_kb: number;
    available_ram_slots: number;
    optical_drive: string;
    nic_speed: string;
    smc_version: string;
    ble_capable: boolean;
    sip_status: string;
    gatekeeper_status: string;
    xprotect_version: string;
    institutional_recovery_key: string;
    disk_encryption_configuration: string;
    filevault2_users: string[];
    storage: Storage[];
    mapped_printers: Printer[];
  };
}

export interface MobileDevice {
  id: number;
  name: string;
  device_name: string;
  udid: string;
  serial_number: string;
  phone_number: string;
  wifi_mac_address: string;
  managed: boolean;
  supervised: boolean;
  model: string;
  model_identifier: string;
  model_display: string;
  username: string;
}

export interface AccountsResponse {
  accounts: AdminsAndGroups;
}

export interface AdminResponse {
  account: Admin;
}

export interface GroupResponse {
  group: Group;
}

export interface UsersResponse {
  users: User[];
}

export interface UserResponse {
  user: User;
}

export interface ApplicationDetailResponse {
  computer_applications: ApplicationDetail;
}

export interface ComputerResponse {
  computers: Computer[];
}

export interface ComputerDetailResponse {
  computer: ComputerDetail;
}

export interface MobileDevicesResponse {
  mobile_devices: MobileDevice[];
}

export interface JamfDataModel {
  users: User[];
  mobileDevices: MobileDevice[];
  computers: Computer[];
  computerDetails: ComputerDetail[];
  admins: Admin[];
  groups: Group[];
}

export enum Method {
  GET = "get",
  POST = "post",
}
