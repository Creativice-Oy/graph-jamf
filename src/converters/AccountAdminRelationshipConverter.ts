import {
  ACCOUNT_ADMIN_RELATIONSHIP_CLASS,
  ACCOUNT_ADMIN_RELATIONSHIP_TYPE,
  ACCOUNT_ENTITY_TYPE,
  AccountAdminRelationship,
  ADMIN_ENTITY_TYPE,
} from "../jupiterone/entities";
import { Account, Admin } from "../types";
import { generateEntityKey, generateRelationKey } from "../utils/generateKey";

export function createAccountAdminRelationships(
  account: Account,
  admins: Admin[],
): AccountAdminRelationship[] {
  const defaultValue: AccountAdminRelationship[] = [];

  return admins.reduce((acc, admin) => {
    const parentKey = generateEntityKey(ACCOUNT_ENTITY_TYPE, account.id);
    const childKey = generateEntityKey(ADMIN_ENTITY_TYPE, admin.id);
    const relationKey = generateRelationKey(
      parentKey,
      ACCOUNT_ADMIN_RELATIONSHIP_TYPE,
      childKey,
    );

    const relationship: AccountAdminRelationship = {
      _class: ACCOUNT_ADMIN_RELATIONSHIP_CLASS,
      _fromEntityKey: parentKey,
      _key: relationKey,
      _type: ACCOUNT_ADMIN_RELATIONSHIP_TYPE,
      _toEntityKey: childKey,
    };

    return [...acc, relationship];
  }, defaultValue);
}
