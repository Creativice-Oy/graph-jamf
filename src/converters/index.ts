export { createUserEntities } from "./UserEntityConverter";
export { createMobileDeviceEntities } from "./MobileDeviceEntityConverter";
export {
  createUserDeviceRelationships,
} from "./UserDeviceRelationshipConverter";
export { createAccountEntity } from "./AccountEntityConverter";
export {
  createAccountUserRelationships,
} from "./AccountUserRelationshipConverter";
export { createComputerEntities } from "./ComputerEntityConverter";
export {
  createUserComputerRelationships,
} from "./UserComputerRelationshipConverter";
