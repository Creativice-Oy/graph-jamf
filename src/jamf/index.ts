import { Account, MobileDevice, User } from "../types";
import fetchJamfData from "./fetchJamfData";
import JamfClient from "./JamfClient";
import { JamfDataModel } from "./types";

export {
  JamfClient,
  fetchJamfData,
  Account,
  User,
  MobileDevice,
  JamfDataModel,
};
