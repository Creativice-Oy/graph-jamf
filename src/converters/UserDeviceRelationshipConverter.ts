import { User } from "../jamf";
import {
  MOBILE_DEVICE_ENTITY_TYPE,
  USER_DEVICE_RELATIONSHIP_CLASS,
  USER_DEVICE_RELATIONSHIP_TYPE,
  USER_ENTITY_TYPE,
  UserDeviceRelationship,
} from "../jupiterone/entities";
import { generateEntityKey, generateRelationKey } from "../utils/generateKey";

export function createUserDeviceRelationships(
  users: User[],
): UserDeviceRelationship[] {
  const defaultValue: UserDeviceRelationship[] = [];

  return users.reduce((acc, user) => {
    if (isMobileDevice(user)) {
      return acc;
    }

    const parentKey = generateEntityKey(USER_ENTITY_TYPE, user.id);
    const childKey = generateEntityKey(
      MOBILE_DEVICE_ENTITY_TYPE,
      user.links!.mobile_devices!.mobile_device!.id,
    );
    const relationKey = generateRelationKey(
      parentKey,
      USER_DEVICE_RELATIONSHIP_CLASS,
      childKey,
    );

    const relationship: UserDeviceRelationship = {
      _class: USER_DEVICE_RELATIONSHIP_CLASS,
      _type: USER_DEVICE_RELATIONSHIP_TYPE,
      _fromEntityKey: parentKey,
      _key: relationKey,
      _toEntityKey: childKey,
    };

    return [...acc, relationship];
  }, defaultValue);
}

function isMobileDevice(user: User) {
  return !(
    user.links &&
    user.links.mobile_devices &&
    user.links.mobile_devices.mobile_device
  );
}
