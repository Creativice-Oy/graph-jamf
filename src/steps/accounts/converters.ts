import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { Account } from '../../jamf/types';
import { generateEntityKey } from '../../util/generateKey';
import { Entities } from '../constants';

export function getAccountEntityKey(accountId: string) {
  return generateEntityKey(Entities.ACCOUNT._type, accountId);
}

export function createAccountEntity(data: Account) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.ACCOUNT._class,
        _type: Entities.ACCOUNT._type,
        _key: getAccountEntityKey(data.id),
        displayName: data.name,
        name: data.name,
      },
    },
  });
}
