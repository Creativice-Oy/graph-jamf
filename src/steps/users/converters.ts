import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { Admin, User } from '../../jamf/types';
import { generateEntityKey } from '../../util/generateKey';
import { skippedRawDataSource } from '../../util/graphObject';
import { Entities } from '../constants';

export function createAdminEntity(data: Admin) {
  return createIntegrationEntity({
    entityData: {
      source: skippedRawDataSource,
      assign: {
        _class: Entities.USER_ADMIN._class,
        _type: Entities.USER_ADMIN._type,
        _key: generateEntityKey(Entities.USER_ADMIN._type, data.id),
        id: data.id.toString(),
        username: data.name,
        displayName: data.name,
        name: data.name,
        admin: true,
        directoryUser: data.directory_user,
        fullName: data.full_name,
        email: data.email?.toLowerCase() || undefined,
        emailAddress: data.email_address?.toLowerCase(),
        active: data.enabled === 'Enabled',
        enabled: data.enabled,
        forcePasswordChange: data.force_password_change,
        accessLevel: data.access_level,
        privilegeSet: data.privilege_set,
        permissions:
          data.privilege_set === 'Administrator'
            ? [data.privilege_set]
            : undefined,
      },
    },
  });
}

export function getUserOperatingSystems(data: User) {
  const os: string[] = [];

  if (data.links) {
    if (data.links.mobile_devices.length > 0) {
      os.push('iOS');
    }

    if (data.links.computers.length > 0) {
      os.push('macOS');
    }
  }

  return os;
}

export function createDeviceUserEntity(data: User) {
  return createIntegrationEntity({
    entityData: {
      source: skippedRawDataSource,
      assign: {
        _class: Entities.DEVICE_USER._class,
        _type: Entities.DEVICE_USER._type,
        _key: generateEntityKey(Entities.DEVICE_USER._type, data.id),
        id: data.id.toString(),
        displayName: data.name,
        name: data.name,
        username: data.name,
        fullName: data.full_name,
        // Jamf device users are always considered active. Jamf _admins_ can
        // be explicitly disabled
        active: true,
        email: data.email?.toLowerCase() || undefined,
        emailAddress: data.email_address?.toLowerCase(),
        phoneNumber: data.phone_number,
        position: data.position,
        enableCustomPhotoUrl: data.enable_custom_photo_url,
        customPhotoUrl: data.custom_photo_url,
        ldapServer: data.ldap_server && data.ldap_server.name,
        totalVppCodeCount: data.links && data.links.total_vpp_code_count,
        os: getUserOperatingSystems(data),
      },
    },
  });
}
