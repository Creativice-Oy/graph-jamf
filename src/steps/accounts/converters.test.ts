import { Account } from '../../jamf/types';
import { createAccountEntity } from './converters';

describe('#createAccountEntity', () => {
  test('should convert data', () => {
    const account: Account = {
      id: 'test-id',
      name: 'test-name',
    };

    expect(createAccountEntity(account)).toMatchSnapshot();
  });
});
