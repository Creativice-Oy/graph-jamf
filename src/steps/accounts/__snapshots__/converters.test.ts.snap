// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createAccountEntity should convert data 1`] = `
Object {
  "_class": Array [
    "Account",
  ],
  "_key": "jamf_account_test-id",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "id": "test-id",
        "name": "test-name",
      },
    },
  ],
  "_type": "jamf_account",
  "createdOn": undefined,
  "displayName": "test-name",
  "id": "test-id",
  "name": "test-name",
}
`;
